### PR TITLE
wifi: Extend regulatory information

### PR DESCRIPTION
**Pull request by ajayparida**
Update regulaory related info like channels supported, dfs support  etc.

### CODE REVIEW

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 1]
```diff
@@ -324,12 +341,17 @@ int nrf_wifi_reg_domain(const struct device *dev, struct wifi_reg_domain *reg_do
 			goto err;
 		}
 	} else if (reg_domain->oper == WIFI_MGMT_GET) {
+
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/src/fmac_main.c:344]: Unnecessary blank line.

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 2]
```diff
@@ -324,12 +341,17 @@ int nrf_wifi_reg_domain(const struct device *dev, struct wifi_reg_domain *reg_do
...
+		rpu_ctx_zep->reg_chnl_info = reg_domain->chnl_info;
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/src/fmac_main.c:345]: I guess a check is necessary to ensure that `reg_domain->chnl_info` is valid ?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c (hunk 2 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/src/fmac_main.c:344]: chnl_info is static memory, so no need to check.

**Review by sachinthegreen — APPROVED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 3]
```diff
@@ -324,12 +341,16 @@ int nrf_wifi_reg_domain(const struct device *dev, struct wifi_reg_domain *reg_do
 			goto err;
 		}
 	} else if (reg_domain->oper == WIFI_MGMT_GET) {
+		rpu_ctx_zep->reg_chan_info = reg_domain->chan_info;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/fmac_main.c:344]: Who is allocating memory for `chan_info`?

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 4]
```diff
@@ -284,6 +287,20 @@ void nrf_wifi_event_get_reg_zep(void *vif_ctx,
 	memcpy(&fmac_dev_ctx->alpha2,
 		   &get_reg_event->nrf_wifi_alpha2,
 		   sizeof(get_reg_event->nrf_wifi_alpha2));
+
+	rpu_ctx_zep->reg_chans_count = get_reg_event->num_channels;
+
+	for (i = 0; i < rpu_ctx_zep->reg_chans_count; i++) {
+		chan_info = &(rpu_ctx_zep->reg_chan_info)[i];
```
[the remark this answers is not in the document]
Comment (ajayparida) [reply drivers/wifi/nrf700x/src/fmac_main.c:294]: Memory resrved and assigned at shell/application . 
It's assigned to driver pointer during command processing 
rpu_ctx_zep->reg_chan_info = reg_domain->chan_info.

**Comment by ajayparida** [comment]
> The commits should be reordered

Updated

File: drivers/wifi/nrf700x/src/fmac_main.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (krish2718) [reply drivers/wifi/nrf700x/src/fmac_main.c:294]: But this isn't thread safe.

**Review by rado17 — APPROVED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 5]
```diff
@@ -281,9 +281,21 @@ void nrf_wifi_event_get_reg_zep(void *vif_ctx,
 	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
 	fmac_dev_ctx = rpu_ctx_zep->rpu_ctx;
 
+	if (fmac_dev_ctx->alpha2_valid) {
+		LOG_ERR("%s: Get regulatory receivedd when req was not sent!", __func__);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/src/fmac_main.c:285]:
```suggestion
		LOG_ERR("%s: Got regulatory information when request was not sent!", __func__);
```

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 6]
```diff
@@ -281,9 +281,21 @@ void nrf_wifi_event_get_reg_zep(void *vif_ctx,
 	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
 	fmac_dev_ctx = rpu_ctx_zep->rpu_ctx;
 
+	if (fmac_dev_ctx->alpha2_valid) {
+		LOG_ERR("%s: Got regulatoryinformation when request was not sent!", __func__);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/fmac_main.c:285]:
```suggestion
		LOG_ERR("%s: Unsolicited regulatory get!", __func__);
```

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 7]
```diff
@@ -281,9 +281,21 @@ void nrf_wifi_event_get_reg_zep(void *vif_ctx,
 	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
 	fmac_dev_ctx = rpu_ctx_zep->rpu_ctx;
 
+	if (fmac_dev_ctx->alpha2_valid) {
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/fmac_main.c:284]: This isn't thread-safe, e.g., calling GET_REG from different threads. The original code for alpha2 itself isn't, so, need to fix that and then the same mechanism can be re-used.

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 8]
```diff
@@ -324,12 +338,33 @@ int nrf_wifi_reg_domain(const struct device *dev, struct wifi_reg_domain *reg_do
 			goto err;
 		}
 	} else if (reg_domain->oper == WIFI_MGMT_GET) {
+
+		if (!reg_domain->chan_info)	{
+			LOG_ERR("%s: No memory allocated for regulatory info\n", __func__);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/fmac_main.c:343]:
```suggestion
			LOG_ERR("%s: Invalid regulatory info (NULL)\n", __func__);
```

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 9]
```diff
@@ -324,12 +338,33 @@ int nrf_wifi_reg_domain(const struct device *dev, struct wifi_reg_domain *reg_do
...
 			LOG_ERR("%s: Failed to get regulatory domain", __func__);
 			goto err;
 		}
 		memcpy(reg_domain->country_code, reg_domain_info.alpha2, WIFI_COUNTRY_CODE_LEN);
+
+		reg_domain->num_channels = reg_domain_info.reg_chan_count;
+		for (chan_idx = 0; chan_idx < reg_domain_info.reg_chan_count; chan_idx++) {
+			chan_info = &(reg_domain->chan_info[chan_idx]);
+			chan_info->center_frequency =
+				reg_domain_info.reg_chan_info[chan_idx].center_frequency;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/fmac_main.c:358]: `reg_domain_info.reg_chan_info[chan_idx]` storing in an intermediate variable will solve the indent issue.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c (hunk 7 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/src/fmac_main.c:285]: Added locking meachanism to sync cmd/event.

**Review by krish2718 — APPROVED** [review]

**Review by sachinthegreen — APPROVED** [review]

**Pull request merged**